### ISSUE TITLE
spUtil

### ISSUE THREAD

**Issue by arthur-cheung** [service-portal/documentation#42]
Hi,

We've been playing around with the service portal widgets and noticed that some existing widgets pulls in the dependency "spUtil" (e.g. "Simple List" widget).

We noticed it contained the following:

- function: update
- function: refresh
- function: triggerTableUIAction
- function: get
- function: getHeaders
- function: getWidgetURL
- function: setBreadCrumb
- function: setSearchPage
- function: addTrivialMessage
- function: addInfoMessage
- function: addErrorMessage
- function: getURL
- function: getHost
- function: scrollTo
- function: getAccelerator
- function: recordWatch

Just wondering if there's any documentation on this. We've searched the Helsinki docs and found a brief description on the [Record Watch functionality](https://docs.servicenow.com/bundle/helsinki-servicenow-platform/page/build/service-portal/concept/c_RecordWatch.html), but would like to know if there's something more comprehensive.

Specifically, is this something that we can use to interact with the server (where you would traditionally use GlideAjax). Is it useful for anything more than record watch (e.g. retrieving and updating tables)?

Also, we noticed $sp in this documentation and it seems both might be of similar nature. Would you be able to explain the difference, and which should be used when?

Cheers,
Arthur

**Comment by enojardi** [comment]
@iamwill @patrickw14  any comments on the difference between $sp and spUtil and what's best practice?

**Comment by CapaJC** [comment]
$sp is an alias for a server-side Java object, available to server scripts. spUtil is client-side (Angular) and available only to the Controller in widgets. 

--J

On Jul 8, 2016, at 2:55 AM, Jonatan Jardi <notifications@github.com> wrote:

@iamwill @patrickw14 any comments on the difference between $sp and spUtil and what's best practice?

**Comment by nathanfirth** [comment]
@CapaJC @enojardi There is also a client side service called spUtil that has a number of useful methods such as: update, refresh, getURL, scrollTo, etc. I have not seen it documented, but being client side you can find the source in Dev Tools.

**Comment by CapaJC** [comment]
No. So we use it on the server, then pass what we need from it to the client. 

--J


On Jul 13, 2016, at 1:48 AM, klima <notifications@github.com> wrote:

In server-side script we have method "getCatalogItem". There is something like this on client-side?

**Comment by SparrowAvanade** [comment]
$sp seems to be the GlideSPScriptable (https://developer.servicenow.com/app.do#!/api_doc?v=helsinki&type=server&scoped=true&to=class__scoped_glidespscriptable__helsinki) Server API.

spUtil is an Angular Service in the Service Portal's "sn.$sp" Angular Module.  You can look at the source (documentation by code) of this entire module yourself by browsing to https://[SN_INSTANCE/scripts/js_includes_sp.jsx on your instance.
